File: guides/micronaut-microstream-persistence/java/src/main/java/example/micronaut/FruitRepository.java
package example.micronaut;

import io.micronaut.core.annotation.NonNull;
import io.micronaut.core.annotation.Nullable;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import java.util.Collection;
import javax.validation.Valid;

interface FruitRepository {

    @NonNull
    Collection<Fruit> list();

    @NonNull
    Fruit create(@NonNull @NotNull @Valid FruitCommand fruit) // <1>
            throws FruitDuplicateException;

    @Nullable
    Fruit update(@NonNull @NotNull @Valid FruitCommand fruit); // <1>

    @Nullable
    Fruit find(@NonNull @NotBlank String name);

    void delete(@NonNull @NotNull @Valid FruitCommand fruit); // <1>
}

File: guides/micronaut-microstream-persistence/kotlin/src/main/kotlin/example/micronaut/FruitRepository.kt
package example.micronaut

internal interface FruitRepository {

    fun list(): Collection<Fruit>

    fun create(fruit: FruitCommand): Fruit

    fun update(fruit: FruitCommand): Fruit?

    fun find(name: String): Fruit?

    fun delete(fruit: FruitCommand)
}